File: src/main/java/com/epri/dss/control/impl/ControlElemImpl.java
package com.epri.dss.control.impl;

import com.epri.dss.common.Bus;
import com.epri.dss.common.CktElement;
import com.epri.dss.common.DSSClass;
import com.epri.dss.common.impl.DSSCktElement;
import com.epri.dss.common.impl.DSSClassDefs;
import com.epri.dss.common.impl.DSSGlobals;
import com.epri.dss.control.ControlElem;

public class ControlElemImpl extends DSSCktElement implements ControlElem {

	private CktElement controlledElement;

	protected String elementName;
	protected int elementTerminal;
	/* If different than terminal */
	protected String controlledBusName;
	protected Bus controlledBus;
	protected String monitorVariable;
	protected int monitorVarIndex;
	protected double timeDelay, dblTraceParameter;
	protected boolean showEventLog;

	public ControlElemImpl(DSSClass parClass) {
		super(parClass);

		objType = DSSClassDefs.CTRL_ELEMENT;
		dblTraceParameter = 0.0;
		monitorVariable = "";
		monitorVarIndex = 0;
		controlledElement = null;
		showEventLog = true;
	}

	/**
	 * Do the action that is pending from last sample.
	 */
	public void doPendingAction(int code, int proxyHdl) {
		DSSGlobals.doSimpleMsg("Programming Error: Reached base class for doPendingAction."+DSSGlobals.CRLF+"Device: "+getDSSClassName()+"."+getName(), 460);
	}

	public void reset() {
		DSSGlobals.doSimpleMsg("Programming Error: Reached base class for reset."+DSSGlobals.CRLF+"Device: "+getDSSClassName()+"."+getName(), 461);
	}

	/**
	 * Sample control quantities and set action times in control queue.
	 */
	public void sample() {
		DSSGlobals.doSimpleMsg("Programming Error: Reached base class for sample."+DSSGlobals.CRLF+"Device: "+getDSSClassName()+"."+getName(), 462);
	}

	public void setControlledElement(CktElement value) {
		try {
			// check for reassignment
			if (controlledElement != null)
				controlledElement.setHasControl(false);
		} finally {
			controlledElement = value;
			if (controlledElement != null) {
				controlledElement.setHasControl(true);
				controlledElement.setControlElement(this);
			}
		}
	}

	public CktElement getControlledElement() {
		return controlledElement;
	}

	public String getElementName() {
		return elementName;
	}

	public void setElementName(String name) {
		elementName = name;
	}

	public int getElementTerminal() {
		return elementTerminal;
	}

	public void setElementTerminal(int terminal) {
		elementTerminal = terminal;
	}

	public String getControlledBusName() {
		return controlledBusName;
	}

	public void setControlledBusName(String name) {
		controlledBusName = name;
	}

	public Bus getControlledBus() {
		return controlledBus;
	}

	public void setControlledBus(Bus bus) {
		controlledBus = bus;
	}

	public String getMonitorVariable() {
		return monitorVariable;
	}

	public void setMonitorVariable(String variable) {
		monitorVariable = variable;
	}

	public int getMonitorVarIndex() {
		return monitorVarIndex;
	}

	public void setMonitorVarIndex(int index) {
		monitorVarIndex = index;
	}

	public double getTimeDelay() {
		return timeDelay;
	}

	public void setTimeDelay(double delay) {
		timeDelay = delay;
	}

	public double getDblTraceParameter() {
		return dblTraceParameter;
	}

	public void setDblTraceParameter(double parameter) {
		dblTraceParameter = parameter;
	}

	public boolean isShowEventLog() {
		return showEventLog;
	}

	public void setShowEventLog(boolean showEventLog) {
		this.showEventLog = showEventLog;
	}

}

File: pydss/control/ControlElemImpl.py
from dss.control.ControlElem import (ControlElem,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.common.impl.DSSClassDefs import (DSSClassDefs,)
from dss.common.impl.DSSCktElement import (DSSCktElement,)


class ControlElemImpl(DSSCktElement, ControlElem):
    controlledElement = None
    elementName = None
    elementTerminal = None
    # If different than terminal
    controlledBusName = None
    controlledBus = None
    monitorVariable = None
    monitorVarIndex = None
    timeDelay = None
    dblTraceParameter = None

    def __init__(self, parClass):
        super(ControlElemImpl, self)(parClass)
        self.objType = DSSClassDefs.CTRL_ELEMENT
        self.dblTraceParameter = 0.0
        self.monitorVariable = ''
        self.monitorVarIndex = 0
        self.controlledElement = None

    def doPendingAction(self, code, proxyHdl):
        """Do the action that is pending from last sample."""
        DSSGlobals.doSimpleMsg('Programming Error: Reached base class for doPendingAction.' + DSSGlobals.CRLF + 'Device: ' + self.getDSSClassName() + '.' + self.getName(), 460)

    def reset(self):
        DSSGlobals.doSimpleMsg('Programming Error: Reached base class for reset.' + DSSGlobals.CRLF + 'Device: ' + self.getDSSClassName() + '.' + self.getName(), 461)

    def sample(self):
        """Sample control quantities and set action times in control queue."""
        DSSGlobals.doSimpleMsg('Programming Error: Reached base class for sample.' + DSSGlobals.CRLF + 'Device: ' + self.getDSSClassName() + '.' + self.getName(), 462)

    def setControlledElement(self, value):
        try:
            if self.controlledElement is not None:
                self.controlledElement.setHasControl(False)
        finally:
            self.controlledElement = value
            if self.controlledElement is not None:
                self.controlledElement.setHasControl(True)
                self.controlledElement.setControlElement(self)
        # check for reassignment

    def getControlledElement(self):
        return self.controlledElement

    def getElementName(self):
        return self.elementName

    def setElementName(self, name):
        self.elementName = name

    def getElementTerminal(self):
        return self.elementTerminal

    def setElementTerminal(self, terminal):
        self.elementTerminal = terminal

    def getControlledBusName(self):
        return self.controlledBusName

    def setControlledBusName(self, name):
        self.controlledBusName = name

    def getControlledBus(self):
        return self.controlledBus

    def setControlledBus(self, bus):
        self.controlledBus = bus

    def getMonitorVariable(self):
        return self.monitorVariable

    def setMonitorVariable(self, variable):
        self.monitorVariable = variable

    def getMonitorVarIndex(self):
        return self.monitorVarIndex

    def setMonitorVarIndex(self, index):
        self.monitorVarIndex = index

    def getTimeDelay(self):
        return self.timeDelay

    def setTimeDelay(self, delay):
        self.timeDelay = delay

    def getDblTraceParameter(self):
        return self.dblTraceParameter

    def setDblTraceParameter(self, parameter):
        self.dblTraceParameter = parameter
